### PR TITLE
Implemented user media manager

### PR DESCRIPTION
**Pull request by jaspervriends**
I've implemented a user media upload manager with the idea of stopping multi-uploads from the same file. The files were already attached to the user, so I did not change anything of that.

**Changelog:**
- Added responsive media manager
- Simplified upload button in the composer
- Added media manager button to the composer
- Added `/fof/uploads` endpoint
- Updated `/fof/upload` endpoint output
- Updated `FileSerializer`
- Added `File` frontend model

**Modal features:**
The media manager is expandable minded. Other extensions could open the `FileManagerModal` modal component and attach custom params to customize their needs.

The file manager is _responsive_, has _drag & drop_ support and automatically selects the uploaded files. After clicking the 'Select files' button, the files will be added with their `bbcode` to the composer.

- `onSelect`: Optional, custom callback, default: `adds bbcode to composer`
- `multiSelect`: Optional, allow multiple selections, default `true`
- `restrictFileType`: Optional, restrict file selection to specific types, default `null`, supported values:
  -  `[array]`
  -  `image`
  - `audio`
  - `video`

**This PR**
Let's focus on the code and improvements/removals:

- General Media Manager code
- Media Management API attributes
- I removed an unused callback from https://github.com/FriendsOfFlarum/upload/blob/master/js/src/forum/components/DragAndDrop.js#L95 and removed the `fof-upload-dropping` and `this.isDropping`.

## Updated composer buttons
image

## Automatically select uploaded files
image

## Example: Restricted to file type `image`
image

## Night mode
image

## Mobile view
image

### CODE REVIEW

**Comment by imorland** [comment]
This really is great work @jaspervriends thank you for your time in creating this feature! Might take a couple of days to go over this, please bear with us :)

**Comment by clarkwinkelmann** [comment]
This looks very nice! I probably also won't be able to review immediately, but I will try to get to it!

How are the image preview managed? Is it the full image? If so, does it load all of the images immediately or only as you scroll down? If a forum allows very large uploads, or someone just has many images, will it not bring the website down by loading all of them?

**Comment by jaspervriends** [comment]
Awesome, take your time :)

> This looks very nice! I probably also won't be able to review immediately, but I will try to get to it!
> 
> How are the image preview managed? Is it the full image? If so, does it load all of the images immediately or only as you scroll down? If a forum allows very large uploads, or someone just has many images, will it not bring the website down by loading all of them?

The files are loaded in batches as usual (20 files each), so it will depend on the overal image size. Currently the images are just using the path of the image. As there are no thumbnails generated by this extention :(

So if someone has any suggestions, let me know. Maybe generate it on-the-fly on a specific endpoint? `fof/uploads/{file-id}/preview`? If so, would we cache that on the server, or just in the browser?

File: resources/less/forum/fileList.less [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+        border-radius: 4px;
+        margin-bottom: 5px;
+        background: @body-bg;
+        background-position: center center;
+        background-size: cover;
+        width: 100%;
+        overflow: hidden;
+        border: 1px solid @control-bg;
+        outline: none;
+        box-shadow: 0 0 0 0px rgba(255, 255, 255, 0);
```
Comment (davwheat) [inline resources/less/forum/fileList.less:66]:
```suggestion
        box-shadow: 0 0 0 0 rgba(255, 255, 255, 0);
```

File: resources/locale/en.yml [hunk 2]
```diff
@@ -92,8 +92,23 @@ fof-upload:
...
       attach: Attach
+      upload: Upload
+      cancel: Cancel
+      media: Media
+      load_more_files: Load more files
+      select_file: Select file|Select files
+
+    file_list:
+      loading: Loading files...
+      empty: There are no files to show.
```
Comment (davwheat) [inline resources/locale/en.yml:107]: Maybe more user-focused, like `You haven't uploaded any files`?

File: resources/locale/en.yml [hunk 3]
```diff
@@ -92,8 +92,23 @@ fof-upload:
...
+      release_to_upload: Release to upload
+      modal_empty_desktop: Your file list is empty, click or drag a file to upload.
```
Comment (davwheat) [inline resources/locale/en.yml:109]: Comma splicing.

```suggestion
      modal_empty_desktop: Your file list is empty. Click or drag a file here to upload.
```

File: resources/locale/en.yml [hunk 4]
```diff
@@ -92,8 +92,23 @@ fof-upload:
...
+      modal_empty_phone: Your file list is empty, select a file to upload.
```
Comment (davwheat) [inline resources/locale/en.yml:110]:
```suggestion
      modal_empty_phone: Your file list is empty. Select a file to upload.
```

**Review by davwheat — COMMENTED** [review]
I'm not in FoF, so I won't make this a proper review, but I just had a few suggestions to point out to do with styling.

File: resources/locale/en.yml (hunk 2 above)
Comment (jaspervriends) [reply resources/locale/en.yml:107]: The idea behind the 'empty' string is that it is for the profile page uploads overview that will only be available for mods or admins (or the user itself). This translatable text will not be shown in the dialog.

The dialog uses the `modal_empty_desktop` and `modal_empty_phone`.

Comment (davwheat) [reply resources/locale/en.yml:107]: Ah gotcha, my bad.

File: resources/less/forum/fileList.less [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+            right: 0;
+            padding: 10px;
+            font-size: 13px;
+            overflow: hidden;
+            text-overflow: ellipsis;
+            white-space: nowrap;
+            -webkit-border-bottom-right-radius: 3px;
+            -webkit-border-bottom-left-radius: 3px;
+            -moz-border-radius-bottomright: 3px;
+            -moz-border-radius-bottomleft: 3px;
```
[the remark this answers is not in the document]
Comment (jaspervriends) [reply resources/less/forum/fileList.less:114]: Good point! Forgot about that 😋 I'll remove them in a future commit

File: resources/less/forum/fileList.less [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+            margin-bottom: 20px;
+            font-size: 35px;
+        }
+    }
+
+    ul {
+        padding: 0px;
+        margin: 0 -7px;
+        
+        display: flex;
```
[the remark this answers is not in the document]
Comment (jaspervriends) [reply resources/less/forum/fileList.less:26]: I considered the `display: grid` too, but I wasn't sure if it was used in Flarum anywhere else, so I went for the safe `flex-box` variant instead. I could change it if needed.

File: resources/less/forum/fileList.less [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+            font-size: 45px;
+        }
+
+        .fof-file-image-preview {
+            position: absolute;
+            top: 50%;
+            left: 50%;
+            max-height: 100%;
+            transform: translate(-50%, -50%);
+            -webkit-user-drag: none;
```
[the remark this answers is not in the document]
Comment (jaspervriends) [reply resources/less/forum/fileList.less:97]: Ah, I did't think about that yet! I'll  update that in a future commit

File: resources/less/forum/fileList.less (hunk 1 above)
Comment (jaspervriends) [reply resources/less/forum/fileList.less:66]: Wouldn't cause any issues indeed, I'll change that

**Comment by luceos** [comment]
I'd love to hear other peoples' opinion about the following. Jasper and I had a discussion about the upload button, previous commits had it replaced by the media manager. Right now there's (yet) another button. Screen space is scarce, especially in the editor.

My suggestion is to allow an admin setting to choose between (upload button, media manager button or both). And then render accordingly. The media manager adds another click to upload.

**Comment by davwheat** [comment]
I'm a fan of that. Maybe it's time to also replace the "Upload" text with a tooltip to match the other buttons in the editor toolbar?

File: src/Api/Controllers/ListUploadsController.php [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+    protected $url;
+
+    public function __construct(UrlGenerator $url)
+    {
+        $this->url = $url;
+    }
+
+    /**
+     * @param \Psr\Http\Message\ServerRequestInterface $request
+     * @param \Tobscure\JsonApi\Document $request
```
Comment (davwheat) [inline src/Api/Controllers/ListUploadsController.php:43]:
```suggestion
     * @param \Tobscure\JsonApi\Document $document
```

File: src/Api/Controllers/UploadController.php [hunk 9]
```diff
@@ -36,12 +38,11 @@ public function __construct(Dispatcher $bus)
 
     /**
      * @param \Psr\Http\Message\ServerRequestInterface $request
+     * @param \Tobscure\JsonApi\Document $request
```
Comment (davwheat) [inline src/Api/Controllers/UploadController.php:41]:
```suggestion
     * @param \Tobscure\JsonApi\Document $document
```

**Comment by jaspervriends** [comment]
@luceos  I think an admin setting would be fine. What would be the default setting then? Both buttons visible? And optionally hide one of the buttons? 

> Maybe it's time to also replace the "Upload" text with a tooltip to match the other buttons in the editor toolbar?
Yup, I already gave the upload button a tooltip :)

**Comment by jaspervriends** [comment]
@luceos @davwheat As requested, I've made some changes and worked on the feedback provided.

I've added an option called `Composer buttons` to the admin dashboard:
image

Which results into: 
- Both buttons visible:
  image
- Upload button only:
  image
- Media manager button only:
  image

If there is anything else, let me know!

**Review by imorland — COMMENTED** [review]
I've been checking out this PR for a couple of days locally, and I must say it works very, very well.

There's a couple of features that I personally feel this needs, and I'd really appreciate the thoughts of others here:

- Ability to remove an upload from the media manager
- Ability to access the media manager from the user profile
- Ability for moderators to access the media of others, with the possibility to remove inappropriate content

I'm quite undecided on if the above should delay merging, or could be follow up changes

**Review by luceos — APPROVED** [review]

File: js/src/admin/components/UploadPage.js [hunk 10]
```diff
@@ -193,6 +201,17 @@ export default class UploadPage extends ExtensionPage {
                                 m('.helpText', app.translator.trans('fof-upload.admin.help_texts.download_templates')),
                                 this.templateOptionsDescriptions(),
                             ]),
+                            m('fieldset', [
+                                m('legend', app.translator.trans('fof-upload.admin.labels.composer_buttons.title')),
+                                m('.helpText', app.translator.trans('fof-upload.admin.help_texts.composer_buttons')),
+                                m('div', [
+                                    Select.component({
```
Comment (davwheat) [inline js/src/admin/components/UploadPage.js:208]: Any chance we could get `className`s put on some of these elements to aid custom styling? :)

File: js/src/forum/components/FileManagerModal.js [hunk 11]
```diff
@@ -0,0 +1,163 @@
...
+                <div className="Modal-content">
+                    <div className="fof-modal-buttons App-backControl">
+                        {UploadButton.component({
+                            uploader: this.uploader,
+                            disabled: app.fileListState.isLoading(),
+                            isMediaUploadButton: true,
+                        })}
+                    </div>
+
+                    <div className={'fof-drag-and-drop'}>
```
Comment (davwheat) [inline js/src/forum/components/FileManagerModal.js:64]: Can these be made consistent with other attributes, just being `className="fof-drag-and-drop"`

File: js/src/admin/components/UploadPage.js [hunk 12]
```diff
@@ -193,6 +201,17 @@ export default class UploadPage extends ExtensionPage {
                                 m('.helpText', app.translator.trans('fof-upload.admin.help_texts.download_templates')),
                                 this.templateOptionsDescriptions(),
                             ]),
+                            m('fieldset', [
```
Comment (davwheat) [inline js/src/admin/components/UploadPage.js:204]: Non-blocking, but we should probably change this stuff to JSX for better readability in another PR.

**Comment by davwheat** [comment]
Not sure on others' opinions, but it might be worth considering `display: grid`. Flarum doesn't use it elsewhere, but some other styles core and this ext use are incompatible with IE11, so using it wouldn't be a major issue, but it'd just make styling far easier.

Up to you at the end of the day -- both work fine!

File: js/src/admin/components/UploadPage.js (hunk 10 above)
Comment (jaspervriends) [reply js/src/admin/components/UploadPage.js:208]: Even in the admin dashboard?

File: js/src/admin/components/UploadPage.js (hunk 12 above)
Comment (imorland) [reply js/src/admin/components/UploadPage.js:204]: Agreed, but as you say, this should be a separate PR to this feature

File: js/src/admin/components/UploadPage.js (hunk 10 above)
Comment (davwheat) [reply js/src/admin/components/UploadPage.js:208]: Some people like to style their backend pages too. Extra classes have never hurt anyone, so I don't see a reason not to provide the ability, even if noone uses it.

**Comment by jaspervriends** [comment]
@imorland I've updated the PR again and added an uploads overview to the user profile page (I agree it is a good addition too). Also, I added a `fof-upload.viewUserUploads` to view uploads from other users for moderation purpose. By default the permission is set to `moderators` in the migration file.

Also, I already added the `fof-upload.deleteUserUploads` to the same migration file for future usage. There is no delete functionality yet, also, I am wondering what would be a good approach for this as there are multiple ways to upload files to different adapters.

When a user sees their own profile, the button label is `My uploads`, when viewing another profile (and having the rights) they'll see `User uploads`.

When you click a file in the user profile page, it will open the file URL.

image

**Comment by imorland** [comment]
Superb @jaspervriends - I'll take a closer look after work tonight!

In regards to deletion, I agree it's a tricky subject. My feeling is that instead of actually deleting the file from disk/bucket/whatever, which in turn could likely impact posts that have it embedded, we should instead 'delete from media manager'.

Perhaps this could be done by adding a column to the `fof-upload-files` table, `null` by default, to act as a flag as to if the file should be included in the media library?

**Comment by imorland** [comment]
Any thoughts on the above @jaspervriends ?

Ideally I'd like to get this merged before we start working on the beta 16 updates for this extension

**Comment by imorland** [comment]
I'll merge this now, @davwheat and I will add the 'delete from media manager' feature as described https://github.com/FriendsOfFlarum/upload/pull/262#issuecomment-777415577

**Pull request merged**